### ISSUE TITLE
Unnecessary "recursive value somevar needs type"

### ISSUE THREAD

**Issue by tanin47** [scala/bug#10811]
I'm using Scala 2.12.4.

```
def test(data: Int): Int = { data + 2 }

def main(): Unit = {
  val result = test(data = 3)
  val data = "some string" + result
}
```

The above fails the compilation with: 

```
[error] /somepathwithline: recursive value result needs type
[error]     val data = "some string" + result
[error]                                ^
[error] one error found
[error] (Test / compileIncremental) Compilation failed
```

But the belows work fine.

```
def test(data: Int): Int = { data + 2 }

def main(): Unit = {
  val result = test(3)  // Please notice that `data =` is removed.
  val data = "some string" + result
}
```

```
def test(data: Int): Int = { data + 2 }

def main(): Unit = {
  val result = test(data = 3) 
  val otherName = "some string" + result  // Please notice that this variable isn't `data`
}
```

**Comment by hrhino** [comment]
This is fixed in 2.13.0-M3, by scala/scala#6092.

**Comment by som-snytt** [comment]
Fixed by https://github.com/scala/scala/pull/6089 for 2.12.4 and `-Xsource:2.13`.

It's too bad that `-Xfatal-warnings` doesn't promote deprecation to error early, so that the compiler could choose the named arg version.
